src/common/primitive_attr_quant.hpp

```diff
@@ -28,175 +28,185 @@
+    int mask_ = INT_MIN;
+    data_type_t data_type_ = data_type::undef;
+    int group_ndims_ = 0;
+    dims_t group_dims_ {};
 };
 
-struct arg_scales_t : public c_compatible {
-    arg_scales_t() = default;
+std::ostream &operator<<(std::ostream &ss, const quant_entry_t &e);
 
-    const runtime_scales_t &get(int arg) const {
-        static const runtime_scales_t default_scales;
+struct scales_t : public c_compatible {
+    scales_t() = default;
+
+    const quant_entry_t &get(int arg) const {
         const auto it = scales_.find(arg);
-        if (it == scales_.end()) return default_scales;
+        if (it == scales_.end()) return default_quant_entry();
         return it->second;
     }
 
-    status_t set(int arg, const runtime_scales_t &scale) {
+    // See `set(...)` comment for `quant_entry_t` for a design choice
+    // explanation.
+    status_t set(int arg, int mask) {
+        return set(arg, mask, default_data_type, 0, {});
+    }
+    status_t set(int arg, int mask, data_type_t data_type, int group_ndims,
+            const dims_t group_dims) {
         if (!check_arg(arg)) return status::invalid_arguments;
-        scales_[arg] = scale;
+        CHECK(scales_[arg].set(mask, data_type, group_ndims, group_dims));
         return status::success;
     }
-
-    bool operator==(const arg_scales_t &rhs) const {
-        return scales_ == rhs.scales_;
-    }
-
-    bool has_default_values(const std::vector<int> &skip_args = {}) const {
-        auto predicate = [](const runtime_scales_t &s) {
-            return s.has_default_values();
-        };
-        return has_default_property(skip_args, predicate);
+    // Use this interface with `default_quant_entry` when need to remove a
+    // specific scale.
+    status_t set(int arg, const quant_entry_t &other) {
+        return scales_[arg].set(other);
     }
 
-    bool has_default_data_type(const std::vector<int> &skip_args = {}) const {
-        auto predicate = [](const runtime_scales_t &s) {
-            return s.has_default_data_type();
-        };
-        return has_default_property(skip_args, predicate);
+    bool has_default_values(const std::vector<int> &supported_args = {}) const {
+        auto predicate
+                = [](const quant_entry_t &s) { return s.has_default_values(); };
+        return has_default_property(supported_args, predicate);
     }
 
-    bool has_default_groups(const std::vector<int> &skip_args = {}) const {
-        auto predicate = [](const runtime_scales_t &s) {
-            return s.has_default_groups();
+    // This interface checks the content of all scales, and allows to ignore
+    // certain arguments.
+    bool has_default_data_type(
+            const std::vector<int> &supported_args = {}) const {
+        auto predicate = [](const quant_entry_t &s) {
+            // Note: `data_type::undef` represents `default_quant_entry`.
+            return utils::one_of(
+                    s.get_data_type(), default_data_type, data_type::undef);
```

Comment:
(nit) maybe add `has_default_data_type` to quant_entry_t ?

Reply:
Discussed offline, `undef` is not the only default data type as there's an interface with just mask which means `f32` is a default data type which is reflected on a scales level.

Added a comment that the call can be removed in the future once each implementation supports data type for scales.